data/tutorials/language/0lg_10_polymorphic_variants.md

```diff
@@ -0,0 +1,713 @@
+---
+id: polymorphic-variants
+title: Polymorphic Variants
+description: |
+  Everything you always wanted to know about polymorphic variants but were afraid to ask
+category: "Language"
+---
+
+# Polymorphic Variants
+
+## Introduction
+
+This tutorial teaches you how to use polymorphic variants. This includes starting to use them, maintaining a project already using them, deciding when to use them or not, and balancing their unique benefits against their drawbacks.
+
+Product types and data types such as `option` and `list` are variants and polymorphic. In this tutorial, they are called _simple variants_ to distinguish them from the _polymorphic variants_ presented here. Simple variants and polymorphic variants are close siblings. Their values are both introduced using labels that may carry data. Both can be recursive and have type parameters. By the way, don't trust a [LLM](https://en.wikipedia.org/wiki/Large_language_model) chatbot if it tells you polymorphic variants are dynamically typed; it is a hallucination. Like simple variants, polymorphic variants are type-checked statically.
+
+However, they are type-checked using different algorithms, which results in a different programming experience. The relationship between value and type (written with the colon symbol `:`) is changed with polymorphic variants. Usually, values are thought of as inhabitants of the type, which is regarded as a set-like thing. Rather, polymorphic variant values should be considered as pieces of data that several functions can accept. Polymorphic variants types are a way to express compatibility relationships between those functions. The approach in this tutorial is to build sense from experience using features of polymorphic variants.
+
```

Comment:
I don't understand what this paragraph is trying to explain.

First, since typecheckers are not the subject on the article, bringing type-checker algorithms only encourage the idea that you can only understand OCaml language if you are a type system expert .

Second, polymorphic variants are using the same type-checking algorithm that the rest of the language.

Third, `None` already belongs to multiple type, there is nothing special at this level with respect to the row type variable of objects or polymorphic variable.

I advocate to completely remove this paragraph.

Reply:
I like the idea of an introduction that would list the practical features of polymorphic variants, from a developer's perspective. I agree it would be friendlier. What do you think of this ?
- Optional declaration
- Tag sharing
- Hash patterns

That creates a different, and probably better, learning track. It implies a large refactoring of the rest of the text, but I agree it's worth the effort.

I'm concerned about the relationship with simple variants. Some newcomers are puzzled by the resemblance-difference with polymorphic variants. Plenty of questions arise in relation to that.